Take in a reference instead of a value

mashedcode opened the pull request:
* Introduce non-backwards compatible changes.
* Take in a references for count_of, remove and remove_times.

The reason for this is that it might be expensive to clone the value.
Also all libstd containers do this.

Maybe this should also be done for insertions because if a duplicate is inserted only the reference is needed but then `where K: Clone` would have to be satisfied. So I'm rather unsure about that.

jmitchell commented on src/multiset.rs:
```diff
@@ -157,28 +157,27 @@ impl<K> HashMultiSet<K> where
+                }
+            }
         }
+        self.elem_counts.remove(val).is_some()
```
Can you please revise this to use `match` with a `mut` reference? The last line is confusing until realizing that the case where only some items are removed returns early.

The cached `size` needs to be updated too, now that #12 is merged.

mashedcode replied:
I don't get this. What `match` on what `mut` reference? `match entry`?
Note that this statement needs to be outside of that extra scope above or the borrow checker will complain.
That's the main reason why this is written in this way.
I agree that this is not very readable.

mashedcode commented:
Your review comment is hard to address. See my comment in the outdated changes above.

mashedcode committed:
Take in a reference instead of a value

* Introduce non-backwards compatible changes.
* Take in a references for count_of, remove and remove_times.

The reason for this is that it might be expensive to clone the value.
Also all libstd containers do this.

jmitchell commented:
Ah, I see what you mean. Thanks.

My only other suggestion after taking a closer look is returning the number of items removed instead of a `bool`. Removing some but not all the requested entries produces a `true` return value, but users may expect `true` to mean `times` entries were removed.

mashedcode commented:
Good idea! Changed to return `usize`.
